c10/Device.h

```diff
@@ -107,12 +107,24 @@ C10_API std::ostream& operator<<(
 namespace std {
 template <>
 struct hash<c10::Device> {
-  size_t operator()(c10::Device device) const noexcept {
-    size_t hash_val = static_cast<size_t>(device.index() + 1);
-    if (device.is_cuda()) {
-      hash_val += 2;
-    }
-    return hash_val;
+  size_t operator()(c10::Device d) const noexcept {
+    // Are you here because this static assert failed?  Make sure you ensure
+    // that the bitmasking code below is updated accordingly!
+    static_assert(sizeof(c10::DeviceType) == 2, "DeviceType is not 16-bit");
+    static_assert(sizeof(c10::DeviceIndex) == 2, "DeviceIndex is not 16-bit");
+    // Note [Hazard when concatenating signed integers]
+    // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    // We must first convert to a same-sized unsigned type, before promoting to
+    // the result type, to prevent sign extension when any of the values is -1.
+    // If sign extension occurs, you'll clobber all of the values in the MSB
+    // half of the resulting integer.
```

Comment:
You could avoid this by doing

    uint32_t bits = static_cast<uint32_t>(d.type()) << 16 ^ static_cast<uint32_t>(d.index());

Note the `^` instead of the `|`. That wouldn't be a concatenation, but it would work for hashing because `^` doesn't lose information even if the MSB is clobbered.
Your solution might be better though because it's easier to understand what it's doing

Reply:
Yeah, it's bad because now you can't unpack the hash in the obvious way (which is what happens in the next patch ;)